lib/straight/blockchain_adapters_dispatcher.rb

```diff
@@ -1,60 +1,67 @@
 
-    STEP = 2
-    attr_reader :list_position, :adapters, :result, :step
+    class AdaptersTimeoutError < TimeoutError; end
+    class AdaptersError < StandardError; end
 
-    def initialize(adapters, &block)
-      @list_position = 0
-      @result        = nil
-      @step          = 0
-      @adapters = adapters
+    TIMEOUT = 60
+    attr_reader :list_position, :adapters, :result, :defer_result, :step, :tasks_parallel_limit
+
+    def initialize(adapters, tasks_parallel_limit: 2, &block)
+      @list_position        = 0
+      @defer_result         = Concurrent::IVar.new
+      @tasks_parallel_limit = tasks_parallel_limit
+      @step                 = 0
+      @adapters             = adapters
       run_requests(block) if block_given?
     end
 
     def get_adapters
-      @step = [adapters.size - @list_position, STEP].min
-      adapters = @adapters[@list_position..@step-1]
+      @step = [adapters.size - @list_position, @tasks_parallel_limit].min
+      adapters = @adapters[@list_position...@list_position+@step]
       @list_position += @step
       adapters
     end
 
     def run_requests(block)
-      threads = []
-      fault_counter = 0
-      get_adapters.each do |adapter| 
-        t = Thread.new do
-          block.call(adapter)
-        end
-        t.abort_on_exception = true
-        threads << t
-      end
-      wait_values(threads)
-    rescue => e
-      raise e if fault_counter == @step && @list_position == @adapters.size
-      if fault_counter == @step 
-        fault_counter = 0
-        retry
-      end
-      fault_counter += 1
+      execute_in_parallel(block, get_adapters)
+      Timeout::timeout(TIMEOUT, AdaptersTimeoutError) {
+        @result = @defer_result.wait.value
+      }
     end
+    
+  private
 
-    def wait_values(threads)
-      threads.each do |t| 
-        Thread.new do
-          if val = t.value
-            @result = val
-            threads.map(&:kill)
-          end
+    def execute_in_parallel(block, adapters)
+      attempts = Concurrent::MVar.new(0)
+      pool = Concurrent::ThreadPoolExecutor.new
+      adapters.each do |adapter|
+        p = Concurrent::Promise.new(executor: pool) { block.call(adapter) }
+        p.on_success do |result|
+          # Straight.logger "[Straight] Adapter #{adapter} has received result: #{result}"
+          @defer_result.set(result)
+          pool.kill
         end
+        p.rescue do |reason|
+          # Straight.logger "[Straight] Adapter #{adapter} got error: #{reason}"
+          raise AdaptersError if finish_iteration?(attempts) && reached_last_adapter?
+          attempts.modify { |v| v+1 }
+          execute_in_parallel(block, get_adapters) if finish_iteration?(attempts)
+        end
+        p.execute
       end
-      loop { break if @result }
-      @result
     end
 
+    def finish_iteration?(attempts)
+      attempts.value == @step
+    end
+
+    def reached_last_adapter?
+      @list_position == @adapter.size
```

Comment:
`@adapter`? I do not see variable with such name. Looks like this is not tested and will raise NoMethodError eventually.